### ISSUE TITLE
Using $1 as a filehandle via open $1, "file" doesn't raise an exception

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7517]
Migrated from [rt.perl.org#31767](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=31767) (status was 'resolved')

Searchable as RT31767$

**Comment by p5pRT** [comment]
### From chris@mail.piyeepress.com

#### Created by chris@mail.piyeepress.com



Open should return false when using $1 as a file handle considering $1 is a read\-only variable&#8203;:

$ perl \-we'open $1\, "/dev/null" or die "open&#8203;: $\!"'

It does not\.  Here's the output after dumping $1&#8203;:

$ perl \-we'open $1\, "/dev/null" or die "open&#8203;: $\!"; use Data&#8203;::Dumper; print Dumper \\$1;'
$VAR1 = \\undef;

Here is when a variable named "$foo" is used as a file handle&#8203;:

$ perl \-we'open $foo\, "/dev/null" or die "open&#8203;: $\!"; use Data&#8203;::Dumper; print Dumper \\$foo;'
$VAR1 = \\\\\*\{'&#8203;::$foo'\};

Attempting to assign to $1 should raise the "Modification of a read\-only value attempted" exception\, right?

I find the fact that open doesn't return false and set $\! unexpected behavior and would classify the behavior as a bug\.

Thanks\,

Chris Angell
chris@&#8203;chrisangell\.com

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

This perlbug was built using Perl v5.8.0 - Wed Sep  3 05:15:53 PDT 2003
It is being executed now by  Perl v5.8.5 - Thu Aug  5 09:52:16 PDT 2004.

Site configuration information for perl v5.8.5:

Configured by chris at Thu Aug  5 09:52:16 PDT 2004.

Summary of my perl5 (revision 5 version 8 subversion 5) configuration:
  Platform:
    osname=openbsd, osvers=3.3, archname=OpenBSD.i386-openbsd-multi
    uname='openbsd mail.piyeepress.com 3.3 bsdkern#0 i386 '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -I/usr/local/include',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='2.95.3 20010125 (prerelease, propolice)', gccosandvers='openbsd3.3'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib
    libs=-lm -lutil -lc
    perllibs=-lm -lutil -lc
    libc=/usr/lib/libc.so.29.0, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=define, ccdlflags=' '
    cccdlflags='-DPIC -fPIC ', lddlflags='-shared -fPIC  -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.5:
    /usr/local/lib/perl5/5.8.5/OpenBSD.i386-openbsd-multi
    /usr/local/lib/perl5/5.8.5
    /usr/local/lib/perl5/site_perl/5.8.5/OpenBSD.i386-openbsd-multi
    /usr/local/lib/perl5/site_perl/5.8.5
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.5:
    HOME=/home/chris
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/bin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/ksh

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Sep 29\, 2004 at 04&#8203;:43&#8203;:50PM \-0000\, Chris Kelly wrote&#8203;:
> Open should return false when using $1 as a file handle considering $1 is a read\-only variable&#8203;:
> 
> $ perl \-we'open $1\, "/dev/null" or die "open&#8203;: $\!"'

Actually\, it should die\, in a similar fashion to the following&#8203;:

&nbsp;    $ perl585 \-we '$1\->\{foo\}=1'
&nbsp;    Modification of a read\-only value attempted at \-e line 1\.

since both involve autovivification of a readonly value\.

The change below to the development version of Perl fixes this\.

Dave\.

\-\- 
The warp engines start playing up a bit\, but seem to sort themselves out
after a while without any intervention from boy genius Wesley Crusher\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#17

Change 23342 by davem@&#8203;davem\-percy on 2004/09/30 20&#8203;:51&#8203;:31

&nbsp;     \[perl \#31767\] open $1\, "file" doesn't raise an exception

Affected files \.\.\.

\.\.\. //depot/perl/pp\.c\#426 edit
\.\.\. //depot/perl/t/io/open\.t\#41 edit

Differences \.\.\.

==== //depot/perl/pp\.c\#426 \(text\) ====

@&#8203;@&#8203; \-159\,6 \+159\,8 @&#8203;@&#8203;
&nbsp;         /\* If this is a 'my' scalar and flag is set then vivify
&nbsp;          \* NI\-S 1999/05/07
&nbsp;          \*/
\+        if \(SvREADONLY\(sv\)\)
\+            Perl\_croak\(aTHX\_ PL\_no\_modify\);
&nbsp;         if \(PL\_op\->op\_private & OPpDEREF\) \{
&nbsp;             char \*name;
&nbsp;             GV \*gv;

==== //depot/perl/t/io/open\.t\#41 \(xtext\) ====

@&#8203;@&#8203; \-12\,7 \+12\,7 @&#8203;@&#8203;
 $Is\_VMS = $^O eq 'VMS';
 $Is\_MacOS = $^O eq 'MacOS';
 
\-plan tests => 107;
\+plan tests => 108;
 
 my $Perl = which\_perl\(\);
 
@&#8203;@&#8203; \-315\,3 \+315\,9 @&#8203;@&#8203;
&nbsp;     'sub f \{ open\(my $fh\, "xxx"\); $fh = "f"; \} f; f;print "ok"'\,
&nbsp;     'ok'\, \{ stderr => 1 \}\,
&nbsp;     '\#29102&#8203;: Crash on assignment to lexical filehandle'\);
\+
\+\# \[perl \#31767\] Using $1 as a filehandle via open $1\, "file" doesn't raise
\+\# an exception
\+
\+eval \{ open $99\, "foo" \};
\+like\($@&#8203;\, qr/Modification of a read\-only value attempted/\, "readonly fh"\);

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu Sep 30 14&#8203;:14&#8203;:22 2004\, davem@&#8203;iabyn\.com wrote&#8203;:
> On Wed\, Sep 29\, 2004 at 04&#8203;:43&#8203;:50PM \-0000\, Chris Kelly wrote&#8203;:
> > Open should return false when using $1 as a file handle considering
> $1 is a read\-only variable&#8203;:
> >
> > $ perl \-we'open $1\, "/dev/null" or die "open&#8203;: $\!"'
> 
> Actually\, it should die\, in a similar fashion to the following&#8203;:
> 
>     $ perl585 \-we '$1\->\{foo\}=1'
>     Modification of a read\-only value attempted at \-e line 1\.
> 
> since both involve autovivification of a readonly value\.
> 
> The change below to the development version of Perl fixes this\.

And causes this bug\, which I’m about to fix shortly&#8203;:

$ perl \-lwe 'close $x; Internals&#8203;::SvREADONLY\($x\,1\);
close $x'
Use of uninitialized value $x in ref\-to\-glob cast at \-e line 1\.
Modification of a read\-only value attempted at \-e line 1\.

I wasn’t trying to modify it\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Thu Sep 30 14&#8203;:14&#8203;:22 2004\, davem@&#8203;iabyn\.com wrote&#8203;:
> On Wed\, Sep 29\, 2004 at 04&#8203;:43&#8203;:50PM \-0000\, Chris Kelly wrote&#8203;:
> > Open should return false when using $1 as a file handle considering
> $1 is a read\-only variable&#8203;:
> >
> > $ perl \-we'open $1\, "/dev/null" or die "open&#8203;: $\!"'
> 
> Actually\, it should die\, in a similar fashion to the following&#8203;:
> 
>     $ perl585 \-we '$1\->\{foo\}=1'
>     Modification of a read\-only value attempted at \-e line 1\.
> 
> since both involve autovivification of a readonly value\.
> 
> The change below to the development version of Perl fixes this\.

And causes this bug\, which I’m about to fix shortly&#8203;:

$ perl \-lwe 'close $x; Internals&#8203;::SvREADONLY\($x\,1\);
close $x'
Use of uninitialized value $x in ref\-to\-glob cast at \-e line 1\.
Modification of a read\-only value attempted at \-e line 1\.

I wasn’t trying to modify it\.

**Comment by p5pRT** [comment]
### From @davidnicol

so symbolic reference is out\, then?

$ perl \-le '"&#8203;:FH&#8203;:"=~/\(\\w\\w\)/ and open $1\, ">&STDOUT"; print FH 99'
99

That's going to stop working?


On Tue\, Aug 23\, 2011 at 4&#8203;:23 PM\, Father Chrysostomos via RT \<
perlbug\-comment@&#8203;perl\.org> wrote&#8203;:

> On Thu Sep 30 14&#8203;:14&#8203;:22 2004\, davem@&#8203;iabyn\.com wrote&#8203;:
> > On Wed\, Sep 29\, 2004 at 04&#8203;:43&#8203;:50PM \-0000\, Chris Kelly wrote&#8203;:
> > > Open should return false when using $1 as a file handle considering
> > $1 is a read\-only variable&#8203;:
> > >
> > > $ perl \-we'open $1\, "/dev/null" or die "open&#8203;: $\!"'
> >
> > Actually\, it should die\, in a similar fashion to the following&#8203;:
> >
> >     $ perl585 \-we '$1\->\{foo\}=1'
> >     Modification of a read\-only value attempted at \-e line 1\.
> >
> > since both involve autovivification of a readonly value\.
> >
> > The change below to the development version of Perl fixes this\.
>
> And causes this bug\, which I’m about to fix shortly&#8203;:
>
> $ perl \-lwe 'close $x; Internals&#8203;::SvREADONLY\($x\,1\);
> close $x'
> Use of uninitialized value $x in ref\-to\-glob cast at \-e line 1\.
> Modification of a read\-only value attempted at \-e line 1\.
>
> I wasn’t trying to modify it\.
>
>


\-\- 
"The tools expect that they have full\, unlimited control of the hardware\."
 \-\- Intel Corporation

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Aug 23 14&#8203;:23&#8203;:57 2011\, sprout wrote&#8203;:
> On Thu Sep 30 14&#8203;:14&#8203;:22 2004\, davem@&#8203;iabyn\.com wrote&#8203;:
> > On Wed\, Sep 29\, 2004 at 04&#8203;:43&#8203;:50PM \-0000\, Chris Kelly wrote&#8203;:
> > > Open should return false when using $1 as a file handle considering
> > $1 is a read\-only variable&#8203;:
> > >
> > > $ perl \-we'open $1\, "/dev/null" or die "open&#8203;: $\!"'
> > 
> > Actually\, it should die\, in a similar fashion to the following&#8203;:
> > 
> >     $ perl585 \-we '$1\->\{foo\}=1'
> >     Modification of a read\-only value attempted at \-e line 1\.
> > 
> > since both involve autovivification of a readonly value\.
> > 
> > The change below to the development version of Perl fixes this\.
> 
> And causes this bug\, which I’m about to fix shortly&#8203;:
> 
> $ perl \-lwe 'close $x; Internals&#8203;::SvREADONLY\($x\,1\);
> close $x'
> Use of uninitialized value $x in ref\-to\-glob cast at \-e line 1\.
> Modification of a read\-only value attempted at \-e line 1\.
> 
> I wasn’t trying to modify it\.

Now fixed with commit ce74145d97\.
